Add release notes for 3.4 release (#9326)
elastic/cloud-on-k8s: docs/release-notes/index.md

Review the changes, fixes, and more in each release of Elastic Cloud on Kubernetes.

## 3.4.0 [elastic-cloud-kubernetes-340-release-notes]

### Release Highlights

#### {{es}} client certificate authentication support

ECK now supports configuring {{es}} to require client certificates for authentication. This allows you to enforce mutual TLS (mTLS) between clients and {{es}}, strengthening security by requiring both the client and server to present valid certificates. Currently, {{es}} and {{product.kibana}} support this feature - {{product.kibana}} can be configured to present client certificates when connecting to {{es}}. Support for the remaining components that connect to {{es}} (Beats, Elastic Agent, APM Server, Logstash, and so on) will follow in future releases. For more details, refer to the [client certificate authentication documentation](docs-content://deploy-manage/security/k8s-es-client-certificate-auth.md).

#### Rolling restarts of {{es}} clusters

> ECK now supports triggering rolling restarts of {{es}} clusters through a new annotation-based mechanism. This enables operators to gracefully restart all nodes in a cluster without manual intervention, useful for applying configuration changes that require a restart. The [rolling restart documentation](docs-content://deploy-manage/deploy/cloud-on-k8s/nodes-orchestration.md#cluster-rolling-restart) provides more details.

Reviewer: The framing here is not correct imo. Configuration changes that require a restart are handled by ECK already. This is more of a troubleshooting mechanism

Author: troubleshooting mechanism it is, changed in fe754604cd6d7cd961fe16ea2250383a7c1e309a

#### Simplified zone awareness configuration

ECK simplifies the configuration of zone awareness for {{es}} clusters, reducing the amount of boilerplate configuration needed to set up topology-aware allocation. For more details, refer to the [zone awareness documentation](docs-content://deploy-manage/deploy/cloud-on-k8s/advanced-elasticsearch-node-scheduling.md#k8s-zone-awareness).

> #### Operator-managed FIPS keystore password support for {{es}}

Reviewer: ```suggestion
#### Automatic password-protected keystore for {{es}} in FIPS mode
```

Author: changed in fe754604cd6d7cd961fe16ea2250383a7c1e309a

> ECK now automatically manages FIPS-compliant keystore passwords for {{es}}. When FIPS mode is enabled in the {{es}} configuration (`xpack.security.fips_mode.enabled: true`), the operator generates, stores, and configures a password-protected keystore — eliminating the need for manual `podTemplate` overrides. This feature activates for {{es}} 9.4.0+ and respects any existing user-provided keystore password configuration. For more details, refer to the [{{es}} FIPS keystore password documentation](docs-content://deploy-manage/deploy/cloud-on-k8s/deploy-fips-compatible-version-of-eck.md#k8s-fips-keystore-password).

Reviewer: "FIPS-compliant" is a loaded term — ECK doesn't guarantee FIPS compliance, it satisfies the Elasticsearch requirement of a password-protected keystore when running in FIPS mode. Suggest rewording to avoid the compliance claim:

```suggestion
#### Operator-managed FIPS keystore password support for {{es}}

ECK now automatically manages a password-protected keystore for {{es}} when FIPS mode is enabled. When `xpack.security.fips_mode.enabled` is set to `true` in the {{es}} configuration, the operator generates, stores, and configures a password-protected keystore — eliminating the need for manual `podTemplate` overrides. This feature activates for {{es}} 9.4.0+ and respects any existing user-provided keystore password configuration. For more details, refer to the [{{es}} FIPS keystore password documentation](docs-content://deploy-manage/deploy/cloud-on-k8s/deploy-fips-compatible-version-of-eck.md#k8s-fips-keystore-password).
```

Author: valid point, changed in fe754604cd6d7cd961fe16ea2250383a7c1e309a

Review the changes, fixes, and more in each release of Elastic Cloud on Kubernetes.

## 3.4.0 [elastic-cloud-kubernetes-340-release-notes]

### Release Highlights

#### {{es}} client certificate authentication support

> ECK now supports configuring {{es}} to require client certificates for authentication. This allows you to enforce mutual TLS (mTLS) between clients and {{es}}, strengthening security by requiring both the client and server to present valid certificates. Currently, {{es}} and {{product.kibana}} support this feature - {{product.kibana}} can be configured to present client certificates when connecting to {{es}}. Support for the remaining components that connect to {{es}} (Beats, Elastic Agent, APM Server, Logstash, and so on) will follow in future releases.

Reviewer 2: ```suggestion
ECK now supports configuring {{es}} to require client certificates for authentication. This allows you to enforce mutual TLS (mTLS) between clients and {{es}}, strengthening security by requiring both the client and server to present valid certificates. Currently, {{es}} and {{product.kibana}} support this feature - {{product.kibana}} can be configured to present client certificates when connecting to {{es}}. Support for the remaining components that connect to {{es}} (Beats, Elastic Agent, APM Server, Logstash, and so on) will follow in future releases. For more details, refer to the [client certificate authentication documentation](docs-content://deploy-manage/security/k8s-es-client-certificate-auth.md).
```

Author: nice point! I completely forgot I can do cross-references to the docs-content repo. changed in dc171208043b755b6d043f04571d1c5384db08dd

#### Rolling restarts of {{es}} clusters

> ECK now supports triggering rolling restarts of {{es}} clusters through a new annotation-based mechanism. This enables operators to gracefully restart all nodes in a cluster without manual intervention, useful for applying configuration changes that require a restart.

Reviewer 2: ```suggestion
ECK now supports triggering rolling restarts of {{es}} clusters through a new annotation-based mechanism. This enables operators to gracefully restart all nodes in a cluster without manual intervention, useful for applying configuration changes that require a restart. The [rolling restart documentation](docs-content://deploy-manage/deploy/cloud-on-k8s/nodes-orchestration.md#cluster-rolling-restart) provides more details.
```

Author: changed in dc171208043b755b6d043f04571d1c5384db08dd

#### Simplified zone awareness configuration

> ECK simplifies the configuration of zone awareness for {{es}} clusters, reducing the amount of boilerplate configuration needed to set up topology-aware allocation.

Reviewer 2: ```suggestion
ECK simplifies the configuration of zone awareness for {{es}} clusters, reducing the amount of boilerplate configuration needed to set up topology-aware allocation. For more details, refer to the [zone awareness documentation](docs-content://deploy-manage/deploy/cloud-on-k8s/advanced-elasticsearch-node-scheduling.md#k8s-zone-awareness).
```

Author: changed in dc171208043b755b6d043f04571d1c5384db08dd

> #### Improved default PVC handling for {{es}}

Reviewer 2: Feels more like a bug fix than a release highlight. It is already listed in the fixes section and, https://github.com/elastic/cloud-on-k8s/pull/9199 itself is labeled `>bug`. 

Also, the PR description calls out a potential breaking impact for users who defined custom PVC volumes in `podTemplate.spec.volumes` with the same name as a default volume (e.g. elasticsearch-data). After upgrading, the operator would try to use those volumes and Kubernetes would reject the StatefulSet update. While likelihood is considered "extremely low", it might be worth a short note on the fixes bullet, or a mention in Breaking Changes?

Author: ok makes sense, changed in 25a2f1ce8aa2cc4a6f5eafb2eb9f5135eb2f77fd

The {{es}} controller now correctly skips adding default volume claim templates whenever a volume with the same name already exists in the pod spec, regardless of volume type. Previously, the operator only skipped defaults for non-PVC volumes (such as `emptyDir` or `hostPath`), which could result in redundant or conflicting volume claim templates when users supplied their own PVC for the same volume name. This change unifies the logic so that user-provided PVCs are respected in the same way as other volume types.

#### Operator-managed FIPS keystore password support for {{es}}

> ECK now automatically manages FIPS-compliant keystore passwords for {{es}}. When FIPS mode is enabled in the {{es}} configuration (`xpack.security.fips_mode.enabled: true`), the operator generates, stores, and configures a password-protected keystore — eliminating the need for manual `podTemplate` overrides. This feature activates for {{es}} 9.4.0+ and respects any existing user-provided keystore password configuration.

Reviewer 2: This feature [doesn't appear to be documented](https://github.com/elastic/docs-content/blob/main/deploy-manage/deploy/cloud-on-k8s/deploy-fips-compatible-version-of-eck.md), not sure if it is expected?

Author: you are 100% right, it wasn't 🙂 I just pushed a commit for it in the docs-content PR and added a reference to that in dc171208043b755b6d043f04571d1c5384db08dd
